src/ciphersuite/mod.rs

```diff
@@ -296,31 +308,96 @@ impl Ciphersuite {
             .unwrap()
     }
 
-    /// Generate a new HPKE key pair and return it.
-    pub(crate) fn derive_hpke_keypair(&self, ikm: &[u8]) -> HPKEKeyPair {
-        self.hpke.derive_key_pair(ikm)
+    /// Derive a new HPKE keypair from a given Secret.
+    pub(crate) fn derive_hpke_keypair(&self, ikm: &Secret) -> HPKEKeyPair {
+        self.hpke.derive_key_pair(ikm.value.as_slice())
     }
 }
 
 impl AeadKey {
-    /// Build a new key for an AEAD from `bytes`.
-    pub(crate) fn from_slice(bytes: &[u8]) -> AeadKey {
+    /// Build a new key for an AEAD from `Secret`.
+    pub(crate) fn from_secret(secret: Secret, aead_mode: AeadMode) -> AeadKey {
+        AeadKey {
+            aead_mode,
+            value: secret.value,
+        }
+    }
+
+    #[cfg(test)]
+    /// Generate a random AEAD Key
+    pub fn new_from_random(aead_mode: AeadMode) -> Self {
         AeadKey {
-            value: bytes.to_vec(),
+            aead_mode,
+            value: aead_key_gen(aead_mode),
         }
     }
 
+    #[cfg(test)]
     /// Get a slice to the key value.
     pub(crate) fn as_slice(&self) -> &[u8] {
         &self.value
     }
+
+    /// Encrypt a payload under the AeadKey given a nonce.
+    pub fn aead_seal(
+        &self,
+        msg: &[u8],
+        aad: &[u8],
+        nonce: &AeadNonce,
+    ) -> Result<Vec<u8>, AEADError> {
+        let (ct, tag) = match aead_encrypt(
+            self.aead_mode,
+            &self.value.as_slice(),
+            msg,
+            &nonce.value,
+            aad,
+        ) {
+            Ok((ct, tag)) => (ct, tag),
+            Err(_) => return Err(AEADError::EncryptionError),
+        };
+        let mut ciphertext = ct;
+        ciphertext.extend_from_slice(&tag);
+        Ok(ciphertext)
+    }
+
+    /// AEAD decrypt `ciphertext` with `key`, `aad`, and `nonce`.
+    pub(crate) fn aead_open(
+        &self,
+        ciphertext: &[u8],
+        aad: &[u8],
+        nonce: &AeadNonce,
+    ) -> Result<Vec<u8>, AEADError> {
+        // TODO: don't hard-code tag bytes
```

Comment:
Any TODO should have an issue. (Not sure if we'll actually do this before there is any PQ scheme.)